aspnetcore/security/authentication/social/additional-claims/samples/2.x/ClaimsSample/Startup.cs

```diff
@@ -30,26 +32,30 @@ public void ConfigureServices(IServiceCollection services)
         {
             services.Configure<CookiePolicyOptions>(options =>
             {
+                // This lambda determines whether user consent for non-essential cookies is needed for a given request.
                 options.CheckConsentNeeded = context => true;
                 options.MinimumSameSitePolicy = SameSiteMode.None;
             });
 
-            services.AddDbContext<AppDbContext>(options => 
+            services.AddDbContext<ApplicationDbContext>(options => 
                 options.UseInMemoryDatabase("InMemoryDb"));
 
-            services.AddIdentity<ApplicationUser, IdentityRole>()
-                .AddEntityFrameworkStores<AppDbContext>()
-                .AddDefaultTokenProviders();
-
             #region snippet_AddGoogle
             services.AddAuthentication().AddGoogle(options =>
             {
                 // Provide the Google Client ID
-                options.ClientId = "XXXXXXXXXXX.apps.googleusercontent.com";
-                // Provide the Google Secret
-                options.ClientSecret = "g4GZ2#...GD5Gg1x";
-                options.Scope.Add("https://www.googleapis.com/auth/plus.login");
-                options.ClaimActions.MapJsonKey(ClaimTypes.Gender, "gender");
+                options.ClientId = "XXXXXXXXXXXXXXX.apps.googleusercontent.com";
+                // Register with User Secrets using:
+                // dotnet user-secrets set "Authentication:Google:ClientId" "{Client ID}"
+
+                // Provide the Google Client Secret
+                options.ClientSecret = "{Client Secret}";
+                // Register with User Secrets using:
+                // dotnet user-secrets set "Authentication:Google:ClientSecret" "{Client Secret}"
+
+                options.Scope.Add("https://www.googleapis.com/auth/userinfo.profile");
+                options.ClaimActions.MapJsonKey(ClaimTypes.GivenName, "givenname");
```

Comment:
Interesting. Doesn't the Google handler already have claim actions to extract those 2 claims?

Reply:
Thanks ... it does. I'd like to use something else.

I'll be free shortly to address this. I'll ping u back in a bit.

Reply:
Actually, it seems the claim names are slightly different: https://github.com/aspnet/AspNetCore/blob/28d44af91a249c9229ba1bbe543898bdbbe4a871/src/Security/Authentication/Google/src/GoogleOptions.cs#L30-L31

Does Google return both claims or was the custom claim action registered in the sample basically ineffective? 😕

Reply:
I need to research this further (and the latest bits more or less match the 2.2 release of that source ... just the Google Plus stuff if a bit different). Commenting out my `MapJsonKey` lines, I still get the names that I used (and not due to an old cookie) ...

```
User Claims
...
http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname
Luke
http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname
Latham
...
```

I'll be free in a bit to dig deeper.

Reply:
For one thing, I didn't set this up correctly ... but it doesn't work anyway yet when I change it to what I think it should be ...

[MapJsonKey](https://docs.microsoft.com/dotnet/api/microsoft.aspnetcore.authentication.claimactioncollectionmapextensions.mapjsonkey) should take care of creating the claim for me.

> Select a top level value from the json user data with the given key name and add it as a Claim. This no-ops if the key is not found or the value is empty.
> 
> ```csharp
> public static void MapJsonKey (
>     this Microsoft.AspNetCore.Authentication.OAuth.Claims.ClaimActionCollection collection, 
>     string claimType, 
>     string jsonKey);
> ```
>
> **claimType** The value to use for Claim.Type when creating a Claim.
>
> **jsonKey** The top level key to look for in the json user data.

Looking at the 2.2 bits :point_right: https://github.com/aspnet/AspNetCore/blob/release/2.2/src/Security/Authentication/Google/src/GoogleOptions.cs

... makes sense. I see ...

```csharp
ClaimActions.MapJsonKey(ClaimTypes.Name, "name");
ClaimActions.MapJsonKey(ClaimTypes.GivenName, "given_name");
ClaimActions.MapJsonKey(ClaimTypes.Surname, "family_name");
ClaimActions.MapJsonKey("urn:google:profile", "link");
ClaimActions.MapJsonKey(ClaimTypes.Email, "email");
```

If I do this, I would expect to see a `Upn` claim that has the value of `name` ...

```csharp
options.ClaimActions.MapJsonKey(ClaimTypes.Upn, "name");
```

That doesn't work. I'm not getting a claim.

I'll need to come back to this over the weekend and see what's what. Marking WIP for now.